src/merkle/merkle_tree.rs

```diff
@@ -115,11 +112,12 @@ impl MerkleTree {
     /// Returns an error if the specified index is not a valid leaf index for this tree.
     pub fn update_leaf(&mut self, index: u64, value: Word) -> Result<(), MerkleError> {
         let depth = self.depth();
-        if index >= 2u64.pow(depth) {
-            return Err(MerkleError::InvalidIndex(depth, index));
+        let node = NodeIndex::new(depth, index);
+        if !node.is_index_valid() {
+            return Err(MerkleError::InvalidIndex(node));
         }
 
-        let mut index = 2usize.pow(depth) + index as usize;
+        let mut index = (node.max_index() + node.index()) as usize;
         self.nodes[index] = value;
```

Comment:
We could change this to:
```Rust
let pos = (index.max_value() + index.value()) as usize;
self.nodes[pos] = value;
```
And then we can update lines 128-127 to use `index.move_up()`. I think it could look like this:
```Rust
for _ in 0..depth {
    let pos = index.value() as usize;
    self.nodes[pos] = Rpo256::merge(&two_nodes[pos]).into();
    index.move_up();
}
```

Reply:
There is a potential problem with the previous `two_nodes`. It is cast as immutable via `two_pairs`, and then we update the values of the nodes as mutable.

To give further context, the compiler can freely move around immutable references, but most likely will not move mutable ones. It's kinda like mutable is [Pin](https://doc.rust-lang.org/std/pin/struct.Pin.html).

In this case, `two_nodes` is a pointer to the "pinned" mutable reference, which is probably fine for most cases since the pointed reference will not move anyway, but it *might*. And if it ever does, this will be undefined behavior as there is no lifetime bound between `two_nodes` and `self.nodes` (if there was, the compiler would reject this operation as we are mutating something that is also borrowed as immutable).

I'll propose something slightly different that achieves the same result without performance impact.